cmd/expand-pool.go

```diff
@@ -0,0 +1,556 @@
+	s.Status = nextPoolExpandStatus(s.Status)
+	return writePoolExpandStats(&SelfPoolExpand{
+		Status:         s.Status,
+		BeforePools:    s.BeforePools,
+		AfterPools:     s.AfterPools,
+		FileMD5:        s.FileMD5,
+		DirHaveRenamed: dirHaveRenamed,
+	})
+}
+
+func nextPoolExpandStatus(ps string) string {
+	switch ps {
+	case "":
+		return PoolExpandStatusWaitForFileChange
+	case PoolExpandStatusWaitForFileChange:
+		return PoolExpandStatusSetEnvToRestart
+	case PoolExpandStatusSetEnvToRestart:
+		return PoolExpandStatusStartDecommission
+	case PoolExpandStatusStartDecommission:
+		return PoolExpandStatusWaitForDecommissionComplete
+	case PoolExpandStatusWaitForDecommissionComplete:
+		return PoolExpandStatusWaitForRenameDataDir
+	case PoolExpandStatusWaitForRenameDataDir:
+		return ""
+	}
+	return ""
+}
+
+func loadPoolExpandStats() (status SelfPoolExpand, err error) {
+	defer logExpandPoolError(err)
+	objectAPI := newObjectLayerFn()
+	if objectAPI == nil {
+		return status, fmt.Errorf("invalid object layer")
+	}
+	z, ok := objectAPI.(*erasureServerPools)
+	if !ok {
+		return status, fmt.Errorf("invalid object layer")
+	}
+	return z.LoadSelfPoolExpand(context.Background())
+}
+
+func getConfigFileInfo() (filemd5 string, pools []string, err error) {
+	if !globalServerCtxt.ExpandPools {
+		return "", nil, fmt.Errorf("pool expand is not enabled")
+	}
+	fileBody, err := os.ReadFile(globalServerCtxt.ConfigPath)
+	if err != nil {
+		return "", nil, err
+	}
+	m5 := md5.New()
+	m5.Write(fileBody)
+	cf := &config.ServerConfig{}
+	dec := yaml.NewDecoder(bytes.NewReader(fileBody))
+	dec.SetStrict(true)
+	if err = dec.Decode(cf); err != nil {
+		return "", nil, err
+	}
+	if len(cf.Pools) != 1 || !cf.EnableExpandPools {
+		return "", nil, fmt.Errorf("invalid config file for expand pool")
+	}
+	return hex.EncodeToString(m5.Sum(nil)), cf.Pools[0], nil
+}
+
+func writePoolExpandStats(poolExpandStatus *SelfPoolExpand) (err error) {
+	defer logExpandPoolError(err)
+	objectAPI := newObjectLayerFn()
+	if objectAPI == nil {
+		return fmt.Errorf("invalid object layer")
+	}
+	z, ok := objectAPI.(*erasureServerPools)
+	if !ok {
+		return fmt.Errorf("invalid object layer")
+	}
+	return z.SaveSelfPoolExpand(context.Background(), *poolExpandStatus)
+}
+
+type renamePath struct {
+	dataPath       string
+	newDataPath    string
+	dataBackupPath string
+}
+
+func initExpandPool(configFile string, ctxt *serverCtxt, pools [][]string) error {
+	if serverDebugLog {
+		log.Infof("[Expand pool]: Starting with: %v", ctxt.Layout)
+	}
+	if !ctxt.ExpandPools {
+		return nil
+	}
+	stat, err := os.Stat(configFile)
+	if err != nil {
+		return err
+	}
+	ltime := stat.ModTime()
+	ctxt.ExpandPoolHandler = func() {
+		if !globalEndpoints.FirstLocal() {
+			return
+		}
+		if serverDebugLog {
+			log.Infof("[Expand pool]: Starting pool expansion")
+		}
+		go func() {
+			ticker := time.NewTicker(time.Second * 10)
+			defer ticker.Stop()
+			for {
+			loop:
+				select {
+				case <-ticker.C:
```

Comment:
ExpandPoolHandler needs to go into its own function, too long and nested to be inside this function.